Yoast/comment-hacks
# QA: create a test setup for the plugin

### QA: add test basics

This commit adds all the necessary prerequisites to the package to allow for adding tests.

#### Composer
* Adds WP Test Utils as a dev dependency.
    This will add PHPUnit, PHPUnit Polyfills and BrainMonkey.
* Adds an `autoload-dev` directive to allow for autoloading test files.
* Adds a `config-platform` directive to ensure that the dependencies in the `composer.lock` file will comply with the minimum supported PHP requirement.
    When different versions of the test utilities are needed, a selective update can be executed in CI, but the locked dependencies should always be compatible with the minimum supported PHP version.
* Adds a `test` script to allow for running the tests using `composer test`.

#### `.gitattributes`
* Exclude files related to the tests from distribution archives.

#### `.gitignore`
* Prevent a `phpunit.xml` overload file from being committed to the repo.
* Prevent the PHPUnit cache file as created by PHPUnit 8/9 from being committed to the repo.

#### `phpunit.xml.dist`
* Add an initial PHPUnit configuration.

This basic setup with WP Test Utils allows for adding both unit tests as well as integration tests.

### QA: add basic setup for integration tests

This commit adds:
* A test bootstrap file for integration tests to be run in a WordPress context.
* A basic test case for integration tests from which all test classes should extend.
* A GitHub Actions workflow to run the integration tests.
    This script will run against all supported WP versions on a mix of supported PHP versions, both with and without multisite.
* A helper script to install the WordPress Core files needed for the tests.
* The test case class to the list of known custom test classes + a minor tweak to the CS rules for the test bootstrap file.

I've elected for the setup for the initial tests to be an integration test setup as that will make adding the first set of tests more straight-forward as WordPress native functions won't need to be mocked.

To run the tests locally, copy the `phpunit.xml.dist` file to a `phpunit.xml` file and add the following, replacing the `value` with the path to a clone of the WordPress core repository you have locally.
```xml
	<php>
		<env name="WP_DEVELOP_DIR" value="path/to/root/of/a/local/WP/clone"/>
	</php>
```

### ~~Tests: add first few tests~~

~~This adds a test class for the `Yoast\WP\Comment\Inc\Clean_Emails` and adds tests for two methods in that class.~~

~~A first test is needed as (depending on the PHPUnit version), the test workflow would fail if there are no tests to run.~~

~~**Note**: in the data provider for the `Clean_Emails::comment_email_headers()` method test, I've added two test cases which IMO highlight bugs in the plugin.~~
~~The tests currently pass, but IMO, the output is not as it should be. Happy to fix the `comment_email_headers()` if I can get confirmation that my assessment that these are bugs is correct.~~

~~@jdevalk 👆🏻 Mind giving an opinion on what I'm pointing out in the above "note" ?~~

---

🆕 I have split the last commit from the original series into the below four commits and added fixes for the identified bugs.

### Tests: add test for Clean_Emails::__construct()

This adds a test class for the `Yoast\WP\Comment\Inc\Clean_Emails` and adds a test for the `__construct()` method in that class.

A first test is needed as (depending on the PHPUnit version), the test workflow would fail if there are no tests to run.

### Tests: add test for Clean_Emails::comment_email_headers()

This commit adds tests for the `comment_email_headers()` method in the `Clean_Emails` class.

The test uses a data provider to test a variation of different cases and safeguards the current functionality.

### Clean_Emails::comment_email_headers(): bug fix for existing headers, but missing content-type

This adds two additional test cases for the `Clean_Emails_Test::test_comment_email_headers()` test method, which would fail without the bug fix.

Both tests essentially test the same thing: a header string is received and contains headers, but no `Content-Type` header.
In that case, the `Content-Type` header needs to be added, but this is not currently handled.

The second part of the commit, adjusts the `Clean_Emails::comment_email_headers()` method to handle these cases correctly.

As each line in a mail header needs to start on a new line and the complete mail header needs to end on a new line character, the second test specifically safeguards this new line handling for the new code.

### Clean_Emails::comment_email_headers(): bug fix for handling unexpected input types

This adds a number of additional test cases for the `Clean_Emails_Test::test_comment_email_headers()` test method, which would fail without the bug fix.

Each of these tests, essentially test the same thing: a `$message_header` input is received, but is not of the type `string`.
While under "normal" circumstances, this shouldn't ever happen, another plugin hooked into the `comment_notification_headers` filter and _doing it wrong_ could lead to this situation.

The new tests document how receiving an unexpected input type will be handled and will safeguard the bug fix which I've put in place for this.
